internal/lefthook/run.go

```diff
@@ -10,6 +56,293 @@ func Run(opts *Options, hookName string, gitArgs []string) error {
+	}
+
+	if len(failList) == 0 {
+		return nil
+	}
+	return errors.New("Have failed script")
+}
+
+func (l *Lefthook) executeCommand(commandName string, gitArgs []string, hook *config.Hook, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if hook.Piped && isPipeBroken {
+		spinner.RestartWithMsg("\n", log.Bold(commandName), log.Brown("(SKIP BY BROKEN PIPE)"))
+		return
+	}
+
+	command := hook.Commands[commandName]
+
+	if command.DoSkip(l.repo.State()) {
+		spinner.RestartWithMsg(successSprint(log.Bold(commandName), log.Brown("(SKIP BY SETTINGS)")))
+		return
+	}
+	if isContainsExcludedTags(command.Tags, hook.ExcludeTags) {
+		spinner.RestartWithMsg(successSprint(log.Bold(commandName), log.Brown("(SKIP BY TAGS)")))
+		return
+	}
+
+	runner := command.GetRunner()
+
+	var files []string
+	filesCommand := hook.Files
+	if command.Files != "" {
+		filesCommand = command.Files
+	}
+
+	if strings.Contains(runner, subStagedFiles) {
+		files, _ = l.repo.StagedFiles()
+	} else if strings.Contains(runner, subFiles) || filesCommand != "" {
+		files, _ = git.FilesByCommand(filesCommand)
+	} else if strings.Contains(runner, pushFiles) {
+		files, _ = l.repo.PushFiles()
+	} else {
+		files, _ = l.repo.AllFiles()
+	}
+
+	log.Debug("\nFiles before filters: \n", files)
+
+	files = FilterGlob(files, command.Glob)
+	files = FilterExclude(files, command.Exclude)
+	files = FilterRelative(files, command.Root)
+
+	log.Debug("\nFiles after filters: \n", files)
+
+	var filesEsc []string
+	for _, fileName := range files {
+		if len(fileName) > 0 {
+			filesEsc = append(filesEsc, shellescape.Quote(fileName))
+		}
+	}
+	files = filesEsc
+
+	log.Debug("Files after escaping: \n", files)
+
+	runner = strings.Replace(runner, pushFiles, strings.Join(files, " "), -1)
+	runner = strings.Replace(runner, subStagedFiles, strings.Join(files, " "), -1)
+	runner = strings.Replace(runner, subAllFiles, strings.Join(files, " "), -1)
+	runner = strings.Replace(runner, subFiles, strings.Join(files, " "), -1)
+	runner = strings.Replace(runner, "{0}", strings.Join(gitArgs, " "), -1)
+	for gitArgIndex, gitArg := range gitArgs {
+		runner = strings.Replace(runner, fmt.Sprintf("{%d}", gitArgIndex+1), gitArg, -1)
+	}
+
+	commandOutput, wait, err := RunCommand(runner, command.Root)
+
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	stageName := fmt.Sprintln(log.Cyan("\n  EXECUTE >"), log.Bold(commandName))
+	if err != nil {
+		failList = append(failList, commandName)
+		setPipeBroken()
+		spinner.RestartWithMsg(stageName, err)
+		return
+	}
+
+	if wait() == nil {
+		spinner.RestartWithMsg(successSprint(stageName, commandOutput.String()))
+
+		okList = append(okList, commandName)
+	} else {
+		spinner.RestartWithMsg(stageName, commandOutput.String())
+
+		failList = append(failList, commandName)
+		setPipeBroken()
+	}
+}
+
+func (l *Lefthook) executeScripts(hookName string, gitArgs []string, sourceDir string, hook *config.Hook, wg *sync.WaitGroup) {
+	sourcePath := filepath.Join(sourceDir, hookName)
```

Comment:
This is the only line where we need hookName and sourceDir. We can call `filepath.Join` on the caller side and pass only sourcePath here, I think it would be more readable.